File: sdk/python/deadroute.py
# sdk/python/deadroute.py
# DeadRoute Python SDK
# Supports: FastAPI, Flask, Django, Starlette
#
# Install: pip install deadroute-python
#
# Quick start (FastAPI):
#   from deadroute import DeadRouteMiddleware
#   app.add_middleware(DeadRouteMiddleware, api_key="dr_live_...", ingest_url="...")
#
# Quick start (Flask):
#   from deadroute import DeadRouteFlask
#   DeadRouteFlask(app, api_key="dr_live_...")

import time
import threading
import queue
import atexit
import re
import os
from typing import Optional, List
from dataclasses import dataclass, field, asdict


@dataclass
class HitPayload:
    method: str
    path: str
    status_code: Optional[int] = None
    duration_ms: Optional[int] = None
    user_agent: Optional[str] = None
    framework: Optional[str] = None


class DeadRouteClient:
    """
    Core DeadRoute client. Manages batching and async flushing.
    Thread-safe, non-blocking — adds < 0.1ms overhead.
    """

    def __init__(
        self,
        api_key: str,
        ingest_url: str = "https://your-app.vercel.app/api/ingest",
        batch_size: int = 50,
        flush_interval: float = 5.0,
        sample_rate: float = 1.0,
        exclude: Optional[List[str]] = None,
        framework: str = "python",
    ):
        self.api_key = api_key
        self.ingest_url = ingest_url
        self.batch_size = batch_size
        self.flush_interval = flush_interval
        self.sample_rate = sample_rate
        self.exclude_patterns = [re.compile(p) for p in (exclude or [])]
        self.framework = framework

        self._queue: queue.Queue = queue.Queue()
        self._stop_event = threading.Event()
        self._worker = threading.Thread(target=self._flush_loop, daemon=True)
        self._worker.start()

        atexit.register(self._shutdown)

    def record(self, hit: HitPayload):
        import random
        if random.random() > self.sample_rate:
            return
        if any(p.search(hit.path) for p in self.exclude_patterns):
            return
        try:
            self._queue.put_nowait(hit)
        except queue.Full:
            pass  # Drop — never block the request

    def _flush_loop(self):
        buffer = []
        while not self._stop_event.is_set():
            deadline = time.time() + self.flush_interval
            while time.time() < deadline:
                try:
                    hit = self._queue.get(timeout=0.1)
                    buffer.append(hit)
                    if len(buffer) >= self.batch_size:
                        self._send(buffer)
                        buffer = []
                except queue.Empty:
                    pass
            if buffer:
                self._send(buffer)
                buffer = []

    def _send(self, hits: list):
        import urllib.request
        import json
        payload = json.dumps({"hits": [
            {k: v for k, v in {
                "method": h.method,
                "path": h.path,
                "statusCode": h.status_code,
                "durationMs": h.duration_ms,
                "userAgent": h.user_agent,
                "framework": h.framework or self.framework,
            }.items() if v is not None}
            for h in hits
        ]}).encode()

        try:
            req = urllib.request.Request(
                self.ingest_url,
                data=payload,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}",
                },
                method="POST",
            )
            urllib.request.urlopen(req, timeout=5)
        except Exception:
            if os.getenv("DEBUG"):
                import traceback
                traceback.print_exc()

    def _shutdown(self):
        self._stop_event.set()
        remaining = []
        try:
            while True:
                remaining.append(self._queue.get_nowait())
        except queue.Empty:
            pass
        if remaining:
            self._send(remaining)


# ─── FastAPI / Starlette ASGI Middleware ──────────────────────────────────────

class DeadRouteMiddleware:
    """
    ASGI middleware for FastAPI and Starlette.

    Usage:
        from fastapi import FastAPI
        from deadroute import DeadRouteMiddleware

        app = FastAPI()
        app.add_middleware(
            DeadRouteMiddleware,
            api_key=os.environ["DEADROUTE_API_KEY"],
            ingest_url="https://your-app.vercel.app/api/ingest",
        )
    """

    def __init__(self, app, api_key: str, ingest_url: str = None, **kwargs):
        self.app = app
        self.client = DeadRouteClient(
            api_key=api_key,
            ingest_url=ingest_url or "https://your-app.vercel.app/api/ingest",
            framework="fastapi",
            **kwargs,
        )

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        start = time.time()
        method = scope.get("method", "GET")
        # Use route pattern if available, else raw path
        path = scope.get("path", "/")
        headers = dict(scope.get("headers", []))
        user_agent = headers.get(b"user-agent", b"").decode("utf-8", errors="ignore")

        status_code = None

        async def send_wrapper(message):
            nonlocal status_code
            if message["type"] == "http.response.start":
                status_code = message.get("status", 200)
            await send(message)

        await self.app(scope, receive, send_wrapper)

        duration_ms = int((time.time() - start) * 1000)
        self.client.record(HitPayload(
            method=method,
            path=path,
            status_code=status_code,
            duration_ms=duration_ms,
            user_agent=user_agent or None,
        ))


# ─── Flask Extension ──────────────────────────────────────────────────────────

class DeadRouteFlask:
    """
    Flask extension for DeadRoute.

    Usage:
        from flask import Flask
        from deadroute import DeadRouteFlask

        app = Flask(__name__)
        DeadRouteFlask(app, api_key=os.environ["DEADROUTE_API_KEY"])
    """

    def __init__(self, app=None, api_key: str = "", **kwargs):
        self.client = None
        self.api_key = api_key
        self.kwargs = kwargs
        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        from flask import request, g
        self.client = DeadRouteClient(
            api_key=self.api_key,
            framework="flask",
            **self.kwargs,
        )

        @app.before_request
        def before_request():
            g._deadroute_start = time.time()

        @app.after_request
        def after_request(response):
            start = getattr(g, "_deadroute_start", time.time())
            self.client.record(HitPayload(
                method=request.method,
                path=request.path,
                status_code=response.status_code,
                duration_ms=int((time.time() - start) * 1000),
                user_agent=request.user_agent.string or None,
            ))
            return response


# ─── Django Middleware ────────────────────────────────────────────────────────

class DeadRouteDjango:
    """
    Django middleware for DeadRoute.

    Add to settings.py MIDDLEWARE list:
        'deadroute.DeadRouteDjango'

    Configure in settings.py:
        DEADROUTE_API_KEY = os.environ["DEADROUTE_API_KEY"]
        DEADROUTE_INGEST_URL = "https://your-app.vercel.app/api/ingest"
    """

    _client: Optional[DeadRouteClient] = None

    def __init__(self, get_response):
        self.get_response = get_response
        from django.conf import settings
        if not self._client:
            DeadRouteDjango._client = DeadRouteClient(
                api_key=getattr(settings, "DEADROUTE_API_KEY", ""),
                ingest_url=getattr(settings, "DEADROUTE_INGEST_URL", "https://your-app.vercel.app/api/ingest"),
                framework="django",
            )

    def __call__(self, request):
        start = time.time()
        response = self.get_response(request)
        duration_ms = int((time.time() - start) * 1000)

        try:
            path = request.resolver_match.route if request.resolver_match else request.path
        except Exception:
            path = request.path

        self._client.record(HitPayload(
            method=request.method,
            path=path,
            status_code=response.status_code,
            duration_ms=duration_ms,
            user_agent=request.META.get("HTTP_USER_AGENT"),
        ))
        return response

File: sdk/ruby/deadroute.rb
# sdk/ruby/deadroute.rb
# DeadRoute Ruby SDK
# Supports: Rails, Sinatra, and any Rack-compatible framework
#
# Install: gem install deadroute-ruby
#
# Quick start (Rails):
#   # config/initializers/deadroute.rb
#   DeadRoute.configure do |c|
#     c.api_key    = ENV["DEADROUTE_API_KEY"]
#     c.ingest_url = ENV["DEADROUTE_INGEST_URL"]
#   end
#
#   # config/application.rb
#   config.middleware.use DeadRoute::Middleware

require "net/http"
require "json"
require "thread"

module DeadRoute
  class Configuration
    attr_accessor :api_key, :ingest_url, :batch_size, :flush_interval,
                  :sample_rate, :exclude, :framework

    def initialize
      @api_key        = ENV["DEADROUTE_API_KEY"] || ""
      @ingest_url     = ENV["DEADROUTE_INGEST_URL"] || "https://your-app.vercel.app/api/ingest"
      @batch_size     = 50
      @flush_interval = 5
      @sample_rate    = 1.0
      @exclude        = []   # Array of Regexp
      @framework      = "rack"
    end
  end

  class << self
    def configuration
      @configuration ||= Configuration.new
    end

    def configure
      yield configuration
    end

    def client
      @client ||= Client.new(configuration)
    end
  end

  # ─── Core Client ─────────────────────────────────────────────────────────────

  class Client
    def initialize(config)
      @config  = config
      @queue   = Queue.new
      @mutex   = Mutex.new
      @running = true

      @worker = Thread.new { flush_loop }
      @worker.abort_on_exception = false

      at_exit { shutdown }
    end

    def record(method:, path:, status_code: nil, duration_ms: nil, user_agent: nil)
      return if rand > @config.sample_rate
      return if @config.exclude.any? { |re| re.match?(path) }

      @queue.push({
        method:      method.to_s.upcase,
        path:        path,
        statusCode:  status_code,
        durationMs:  duration_ms,
        userAgent:   user_agent,
        framework:   @config.framework,
      }.compact)
    end

    private

    def flush_loop
      buffer = []
      while @running
        deadline = Time.now + @config.flush_interval
        while Time.now < deadline
          begin
            hit = @queue.pop(true)   # non-blocking
            buffer << hit
            if buffer.size >= @config.batch_size
              send_hits(buffer)
              buffer = []
            end
          rescue ThreadError
            sleep 0.1
          end
        end
        send_hits(buffer) unless buffer.empty?
        buffer = []
      end
    end

    def send_hits(hits)
      return if hits.empty?
      uri  = URI.parse(@config.ingest_url)
      body = JSON.generate({ hits: hits })

      http             = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl     = uri.scheme == "https"
      http.open_timeout = 3
      http.read_timeout = 5

      request = Net::HTTP::Post.new(uri.request_uri)
      request["Content-Type"]  = "application/json"
      request["Authorization"] = "Bearer #{@config.api_key}"
      request.body             = body

      http.request(request)
    rescue StandardError
      # Never raise — always silent in production
    end

    def shutdown
      @running = false
      remaining = []
      remaining << @queue.pop(true) until @queue.empty? rescue nil
      send_hits(remaining) unless remaining.empty?
    end
  end

  # ─── Rack Middleware ──────────────────────────────────────────────────────────

  class Middleware
    def initialize(app, options = {})
      @app = app
      if options.any?
        DeadRoute.configure do |c|
          c.api_key        = options[:api_key]        if options[:api_key]
          c.ingest_url     = options[:ingest_url]     if options[:ingest_url]
          c.sample_rate    = options[:sample_rate]    if options[:sample_rate]
          c.exclude        = options[:exclude]        if options[:exclude]
          c.framework      = options[:framework]      if options[:framework]
        end
      end
    end

    def call(env)
      start  = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      status, headers, body = @app.call(env)
      duration_ms = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000).round

      request = Rack::Request.new(env)
      path    = route_path(env) || request.path

      DeadRoute.client.record(
        method:      request.request_method,
        path:        path,
        status_code: status,
        duration_ms: duration_ms,
        user_agent:  env["HTTP_USER_AGENT"],
      )

      [status, headers, body]
    end

    private

    # Prefer the Rails route pattern (e.g. /users/:id) over the raw path
    def route_path(env)
      env["action_dispatch.request.path_parameters"]&.then do
        env["action_dispatch.routes"]
          &.recognize_path(env["PATH_INFO"], method: env["REQUEST_METHOD"])
          &.then { nil }  # just fall through; Rails normalizes in rack itself
      end
      env.dig("action_dispatch.route", "path") ||
        env["sinatra.route"]&.split&.last ||
        nil
    rescue StandardError
      nil
    end
  end
end
